File: apps/web-admin/src/types/rig.ts
/**
 * Rig configuration types
 */

// Rig category classification
export type RigCategory = 'bottom' | 'float' | 'lure' | 'fly' | 'surf'

// Rig difficulty levels
export type RigDifficulty = 'easy' | 'medium' | 'hard'

// Component types
export type ComponentType =
  | 'hook'      // 钓钩
  | 'sinker'    // 铅坠
  | 'swivel'    // 转环
  | 'leader'    // 前导线
  | 'float'     // 浮漂
  | 'bead'      // 珠子
  | 'stopper'   // 太空豆
  | 'snap'      // 别针
  | 'ring'      // 连接环
  | 'tube'      // 套管
  | 'other'     // 其他

// Rig specification
export interface RigSpec {
  spec_id: number
  rig_id: number
  spec_name: string
  spec_value: string
  unit?: string
  notes?: string
}

// Rig component
export interface RigComponent {
  component_id: number
  rig_id: number
  component_name: string
  component_type: string
  quantity: number
  size?: string
  position?: number
  notes?: string
}

// Full rig detail
export interface Rig {
  rig_id: number
  name: string
  category: RigCategory
  description?: string
  diagram_url?: string
  difficulty: RigDifficulty
  target_species?: string
  best_conditions?: string
  created_at?: string
  updated_at?: string
  specs: RigSpec[]
  components: RigComponent[]
}

// Rig list item (without nested details)
export interface RigListItem {
  rig_id: number
  name: string
  category: RigCategory
  difficulty: RigDifficulty
  target_species?: string
  diagram_url?: string
  component_count: number
  spec_count: number
  created_at?: string
  updated_at?: string
}

// Paginated list response
export interface RigListResponse {
  total: number
  page: number
  page_size: number
  items: RigListItem[]
}

// Create rig request
export interface RigCreateRequest {
  name: string
  category: RigCategory
  description?: string
  diagram_url?: string
  difficulty?: RigDifficulty
  target_species?: string
  best_conditions?: string
  specs?: RigSpecCreate[]
  components?: RigComponentCreate[]
}

// Update rig request
export interface RigUpdateRequest {
  name?: string
  category?: RigCategory
  description?: string
  diagram_url?: string
  difficulty?: RigDifficulty
  target_species?: string
  best_conditions?: string
}

// Create spec request
export interface RigSpecCreate {
  spec_name: string
  spec_value: string
  unit?: string
  notes?: string
}

// Update spec request
export interface RigSpecUpdate {
  spec_name?: string
  spec_value?: string
  unit?: string
  notes?: string
}

// Create component request
export interface RigComponentCreate {
  component_name: string
  component_type: string
  quantity?: number
  size?: string
  position?: number
  notes?: string
}

// Update component request
export interface RigComponentUpdate {
  component_name?: string
  component_type?: string
  quantity?: number
  size?: string
  position?: number
  notes?: string
}

// Simplified lure type for associations
export interface LureTypeSimple {
  lure_type_id: number
  name: string
  category?: string
}

// Lure type association request
export interface RigLureAssociationRequest {
  lure_type_ids: number[]
}

// Option item for dropdowns
export interface OptionItem {
  value: string
  label: string
  description?: string
  color?: string
  icon?: string
}

// Rig form options response
export interface RigOptionsResponse {
  categories: OptionItem[]
  difficulties: OptionItem[]
  component_types: OptionItem[]
}

// ========== Configuration Constants ==========

// Rig category configuration
export const RIG_CATEGORY_CONFIG: Record<RigCategory, {
  label: string
  icon: string
  color: string
  description: string
}> = {
  bottom: {
    label: '底钓钓组',
    icon: '⚓',
    color: 'volcano',
    description: '适合底层鱼类，如鲤鱼、鲫鱼',
  },
  float: {
    label: '浮漂钓组',
    icon: '🎈',
    color: 'blue',
    description: '适合中上层鱼类，如草鱼、鲢鱼',
  },
  lure: {
    label: '路亚钓组',
    icon: '🎣',
    color: 'green',
    description: '适合掠食性鱼类，如黑鲈、鳜鱼',
  },
  fly: {
    label: '飞蝇钓组',
    icon: '🦋',
    color: 'purple',
    description: '适合溪流鱼类，如虹鳟、马口',
  },
  surf: {
    label: '海钓钓组',
    icon: '🌊',
    color: 'cyan',
    description: '适合海水鱼类，如海鲈、真鲷',
  },
}

// Rig difficulty configuration
export const RIG_DIFFICULTY_CONFIG: Record<RigDifficulty, {
  label: string
  color: string
  description: string
}> = {
  easy: {
    label: '简单',
    color: 'success',
    description: '新手友好，组装简单',
  },
  medium: {
    label: '中等',
    color: 'warning',
    description: '需要一定经验',
  },
  hard: {
    label: '困难',
    color: 'error',
    description: '需要丰富经验和技巧',
  },
}

// Component type options
export const COMPONENT_TYPE_OPTIONS: OptionItem[] = [
  { value: 'hook', label: '钓钩', icon: '🪝', color: 'red' },
  { value: 'sinker', label: '铅坠', icon: '⚫', color: 'default' },
  { value: 'swivel', label: '转环', icon: '🔄', color: 'blue' },
  { value: 'leader', label: '前导线', icon: '〰️', color: 'green' },
  { value: 'float', label: '浮漂', icon: '🎈', color: 'orange' },
  { value: 'bead', label: '珠子', icon: '⚪', color: 'purple' },
  { value: 'stopper', label: '太空豆', icon: '⏹️', color: 'cyan' },
  { value: 'snap', label: '别针', icon: '📎', color: 'gold' },
  { value: 'ring', label: '连接环', icon: '⭕', color: 'lime' },
  { value: 'tube', label: '套管', icon: '🔲', color: 'magenta' },
  { value: 'other', label: '其他', icon: '📦', color: 'default' },
]

// Helper function to get component type config
export const getComponentTypeConfig = (type: string): OptionItem => {
  return COMPONENT_TYPE_OPTIONS.find(opt => opt.value === type)
    || { value: type, label: type, color: 'default' }
}

File: apps/api/schemas/rig.py
"""
Rig configuration API schemas
"""

from typing import Optional, List
from pydantic import BaseModel, Field, ConfigDict
from datetime import datetime
from enum import Enum


class RigCategoryEnum(str, Enum):
    """Rig category classification"""
    BOTTOM = "bottom"       # 底钓钓组
    FLOAT = "float"         # 浮漂钓组
    LURE = "lure"           # 路亚钓组
    FLY = "fly"             # 飞蝇钓组
    SURF = "surf"           # 海钓钓组


class RigDifficultyEnum(str, Enum):
    """Rig difficulty levels"""
    EASY = "easy"
    MEDIUM = "medium"
    HARD = "hard"


# ========== Component Schemas ==========

class RigComponentBase(BaseModel):
    """Base rig component schema"""
    component_name: str = Field(..., min_length=1, max_length=100, description="组件名称")
    component_type: str = Field(..., max_length=50, description="组件类型")
    quantity: int = Field(default=1, ge=1, description="数量")
    size: Optional[str] = Field(None, max_length=50, description="尺寸规格")
    position: Optional[int] = Field(None, ge=0, description="组装位置")
    notes: Optional[str] = Field(None, max_length=500, description="组装说明")


class RigComponentCreate(RigComponentBase):
    """Create rig component"""
    pass


class RigComponentUpdate(BaseModel):
    """Update rig component"""
    component_name: Optional[str] = Field(None, min_length=1, max_length=100)
    component_type: Optional[str] = Field(None, max_length=50)
    quantity: Optional[int] = Field(None, ge=1)
    size: Optional[str] = Field(None, max_length=50)
    position: Optional[int] = Field(None, ge=0)
    notes: Optional[str] = Field(None, max_length=500)


class RigComponentResponse(RigComponentBase):
    """Rig component response"""
    component_id: int
    rig_id: int

    model_config = ConfigDict(from_attributes=True)


# ========== Spec Schemas ==========

class RigSpecBase(BaseModel):
    """Base rig specification schema"""
    spec_name: str = Field(..., min_length=1, max_length=100, description="规格名称")
    spec_value: str = Field(..., max_length=200, description="规格值")
    unit: Optional[str] = Field(None, max_length=20, description="单位")
    notes: Optional[str] = Field(None, max_length=500, description="备注")


class RigSpecCreate(RigSpecBase):
    """Create rig specification"""
    pass


class RigSpecUpdate(BaseModel):
    """Update rig specification"""
    spec_name: Optional[str] = Field(None, min_length=1, max_length=100)
    spec_value: Optional[str] = Field(None, max_length=200)
    unit: Optional[str] = Field(None, max_length=20)
    notes: Optional[str] = Field(None, max_length=500)


class RigSpecResponse(RigSpecBase):
    """Rig specification response"""
    spec_id: int
    rig_id: int

    model_config = ConfigDict(from_attributes=True)


# ========== Rig Schemas ==========

class RigBase(BaseModel):
    """Base rig schema"""
    name: str = Field(..., min_length=1, max_length=100, description="钓组名称")
    category: str = Field(..., description="钓组分类")
    description: Optional[str] = Field(None, max_length=2000, description="详细描述")
    diagram_url: Optional[str] = Field(None, max_length=500, description="示意图URL")
    difficulty: str = Field(default="medium", description="难度等级")
    target_species: Optional[str] = Field(None, max_length=200, description="目标鱼种")
    best_conditions: Optional[str] = Field(None, max_length=1000, description="最佳钓鱼条件")


class RigCreate(RigBase):
    """Create rig - with optional nested specs and components"""
    specs: Optional[List[RigSpecCreate]] = Field(default=None, description="规格参数列表")
    components: Optional[List[RigComponentCreate]] = Field(default=None, description="组件配件列表")


class RigUpdate(BaseModel):
    """Update rig - All fields optional"""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    category: Optional[str] = None
    description: Optional[str] = Field(None, max_length=2000)
    diagram_url: Optional[str] = Field(None, max_length=500)
    difficulty: Optional[str] = None
    target_species: Optional[str] = Field(None, max_length=200)
    best_conditions: Optional[str] = Field(None, max_length=1000)


class RigResponse(RigBase):
    """Rig detail response with nested specs and components"""
    rig_id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    specs: List[RigSpecResponse] = []
    components: List[RigComponentResponse] = []

    model_config = ConfigDict(from_attributes=True)


class RigListItem(BaseModel):
    """Rig list item - simplified without nested details"""
    rig_id: int
    name: str
    category: str
    difficulty: str
    target_species: Optional[str] = None
    diagram_url: Optional[str] = None
    component_count: int = 0
    spec_count: int = 0
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    model_config = ConfigDict(from_attributes=True)


class RigListResponse(BaseModel):
    """Paginated rig list response"""
    total: int
    page: int
    page_size: int
    items: List[RigListItem]


# ========== Lure Type Association ==========

class LureTypeSimple(BaseModel):
    """Simplified lure type for associations"""
    lure_type_id: int
    name: str
    category: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class RigLureAssociationRequest(BaseModel):
    """Request to associate lure types with a rig"""
    lure_type_ids: List[int] = Field(..., min_length=0, description="拟饵类型ID列表")


# ========== Option Response ==========

class OptionItem(BaseModel):
    """Generic option item"""
    value: str
    label: str
    description: Optional[str] = None
    color: Optional[str] = None
    icon: Optional[str] = None


class RigOptionsResponse(BaseModel):
    """Rig form options"""
    categories: List[OptionItem]
    difficulties: List[OptionItem]
    component_types: List[OptionItem]


# ========== Fetch Progress Schemas ==========

class RigFetchProgressItem(BaseModel):
    """钓组采集进度项"""
    name_cn: str
    name_en: Optional[str] = None
    status: str
    rig_id: Optional[int] = None
    error: Optional[str] = None
    updated_at: Optional[str] = None


class RigFetchProgressStats(BaseModel):
    """钓组采集统计"""
    total: int = 0
    pending: int = 0
    fetching: int = 0
    enriching: int = 0
    completed: int = 0
    failed: int = 0
    skipped: int = 0


class RigFetchProgressResponse(BaseModel):
    """钓组采集进度响应"""
    is_running: bool = False
    stats: RigFetchProgressStats
    items: List[RigFetchProgressItem]


class RigFetchActionResponse(BaseModel):
    """钓组采集操作响应"""
    success: bool
    message: str


class RigFetchRetryRequest(BaseModel):
    """钓组采集重试请求"""
    names: Optional[List[str]] = Field(None, description="要重试的钓组名称列表，为空则重试所有失败项")
